README.md
<!-- README.md is generated from README.Rmd. Please edit that file -->

akmedoids
=========

<!-- badges: start -->
<!-- badges: end -->

An R package for analyzing and clustering longitudinal data

### Description

The `akmedoids` package advances the clustering of longitudinal datasets
in order to identify clusters of trajectories with similar long-term
linear trends over time, providing an improved cluster identification as
compared with the classic kmeans algorithm. The package also includes a
set of functions for addressing common data issues, such as missing
entries and outliers, prior to conducting advance longitudinal data
analysis. One of the key objectives of this package is to facilitate
easy replication of a recent paper which examined small area inequality
in the crime drop (Adepeju et al. 2020). Many of the functions provided
in the `akmedoids` package may be applied to longitudinal data in
general.

**For more information and usability, check out details on
[CRAN](https://cran.r-project.org/web/packages/akmedoids/index.html).**

### Installation from `CRAN`

From an R console, type:

    #install.packages("akmedoids")
    library(akmedoids)

    #Other libraries
    library(tidyr)
    library(ggplot2)
    #> Warning: package 'ggplot2' was built under R version 4.0.3
    library(reshape)
    #> Warning: package 'reshape' was built under R version 4.0.3
    library(readr)

To install the development version of the package, type
`remotes::install_github("MAnalytics/akmedoids")`. Please, report any
installation problems in the issues.

### Example usage:

Given a longitudinal datasets, the following is an example of how
`akmedoids` could be used to extract clusters of trajectories with
similar long-term trends over time. We will use a simulated dataset
(named `simulated.rda`) stored in the `data/` directory.

### Generating artificial dataset

Simulating data set which comprised of three clusters with distinct mean
directional change over time. Each group contains 50 trajectories.

    dir.create("input") # create a folder
    #> Warning in dir.create("input"): 'input' already exists

    #function for creating longitudinal noise
    noise_fn <- function(x=3, time){
      rnorm(length(time), mean=0, x)}

    #function for simulating a trajectory group
    sim_group <- function(gr_baseline, sd, time){
      intcp_errors <- rgamma(1, shape=2, scale=sd) #intercept error
      mean_traj = gr_baseline + intcp_errors
      traj = mean_traj + noise_fn(intcp_errors, time)
    }

    #time steps
    t_steps <- c(0:20)

    #increasing group
    i_gr <- NULL
    for(i in seq_len(50)){
      i_gr <- rbind(i_gr,
                    sim_group(gr_baseline=(0.5*t_steps),
                              sd=1, time=t_steps))
    }

    #stable group
    s_gr <- NULL
    for(i in seq_len(50)){
      s_gr <- rbind(s_gr,
                    sim_group(gr_baseline=rep(3,length(t_steps)),
                              sd=1, time=t_steps))
    }

    #decreasing group
    d_gr <- NULL
    for(i in seq_len(50)){
      d_gr <- rbind(d_gr,
                    sim_group(gr_baseline=(10 - (0.5*t_steps)),
                              sd=1, time=t_steps))
    }

    #combine groups
    simulated <- data.frame(rbind(i_gr, s_gr, d_gr))

    #add group label
    simulated <- data.frame(cbind(ID=1:nrow(simulated), simulated))

    colnames(simulated) <- c("ID", 1:(ncol(simulated)-1))

    #save data set
    ##simulated = readr::write_csv(simulated, "input/example-simulated.csv")

### Visualising artificial dataset


    #import already save simulated data
    Import_simulated = read_csv(file="./input/example-simulated.csv")

    #preview the data
    head(Import_simulated)
    #> # A tibble: 6 x 22
    #>      ID    `1`   `2`    `3`    `4`    `5`   `6`   `7`   `8`   `9`  `10`  `11`
    #>   <dbl>  <dbl> <dbl>  <dbl>  <dbl>  <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl>
    #> 1     1 -0.759  1.90  3.02  -0.355  1.63   4.24  4.52  7.93  5.18  6.84  6.09
    #> 2     2  5.52   3.22  4.49   4.61  11.2    5.90 15.7   7.76  7.46  9.76  6.61
    #> 3     3  1.52   1.09  2.05   1.89   0.728  1.27  4.76  3.75  4.58  5.81  5.39
    #> 4     4  3.24   2.41 -0.169  5.59   1.29   3.42  4.59  7.29  5.23  7.25  6.90
    #> 5     5  0.950  2.58  2.58   1.39   3.22   3.28  3.40  6.21  5.71  3.69  7.39
    #> 6     6  2.33   7.88  2.15   1.89   7.38   3.57  7.33 10.4   5.90  5.12 10.3 
    #> # ... with 10 more variables: `12` <dbl>, `13` <dbl>, `14` <dbl>, `15` <dbl>,
    #> #   `16` <dbl>, `17` <dbl>, `18` <dbl>, `19` <dbl>, `20` <dbl>, `21` <dbl>


    #convert wide-format into long
    simulated_long <- melt(t(Import_simulated), id.vars=c("ID"))

     simulated_long <- simulated_long %>%
      dplyr::filter(X1!="ID") %>%
      dplyr::rename(Time=X1, ID=X2)
    # 
    simulated_long <- data.frame(cbind(simulated_long,
                                       Groups= c(rep("Increasing", 50*21),
                                       rep("Stable", 50*21),
                                       rep("Decreasing", 50*21))))

    #re-order levels
    simulated_long$Time <- factor(simulated_long$Time, 
                                    levels = c(1:21))

    ggplot(simulated_long, aes(x = Time, y = value, group=ID, color=Groups)) +
      geom_point(size=0.5) + 
      geom_line() +
      scale_color_manual(values=c('#999999','#E69F00', '#56B4E9')) +
      theme_light()

<img src="man/figures/README-unnamed-chunk-4-1.png" width="100%" />

### Performing clustering using `akmedoids`

Performing clustering analysis using `akmedoids` package.


    output <- akclustr(Import_simulated, id_field=TRUE, verbose = FALSE, k=c(3,12), crit = "Silhouette",
                      quality_plot=TRUE)
    #> `geom_smooth()` using formula 'y ~ x'

<img src="man/figures/README-unnamed-chunk-5-1.png" width="100%" />


    #Remark: The entire output can be printed out to the console by typing `output`

    #To check the optimal cluster size, type:
    output$optimal_k
    #> [1] 3.000046

    #Show quality plot
    #plot(output$QltyPlot)

    #user may decide the examine the plot post clustering....

### Documentation

From an R console type `??akmedoids` for help on the package. The
package page on CRAN is
[here](https://cran.r-project.org/web/packages/akmedoids/index.html),
package reference manual is
[here](https://cran.r-project.org/web/packages/akmedoids/akmedoids.pdf),
package vignette is
[here](https://cran.r-project.org/web/packages/akmedoids/vignettes/akmedoids-vignette.html).

### Support and Contributions:

For support and bug reports send an email to:
<a href="mailto:monsuur2010@yahoo.com" class="email">monsuur2010@yahoo.com</a>
or open an issue [here](https://github.com/MAnalytics/akmedoids/issues).
Code contributions to akmedoids are also very welcome.

### References:

Rousseeuw, P. J. 1987. “Silhouettes: A Graphical Aid to the
Interpretation and Validation of Cluster Analysis.” Journal of
Computational and Applied Mathematics, no. 20: 53–6.
[link](https://www.bibsonomy.org/bibtex/bc0f62c7895f91c787354d03f23da976)

Caliński, T., and J. Harabasz. 1974. “A Dendrite Method for Cluster
Analysis.” Communications in Statistics-Theory and Methods, 3(1): 1–27.
[link](https://www.tandfonline.com/doi/abs/10.1080/03610927408827101)

Adepeju, M., Langton, S. and Bannister, J. 2020. Anchored k-medoids: a
novel adaptation of k-means further refined to measure instability in
the exposure to crime. Journal of Computational Social Science,
(revised).


NEWS.md
---
title: "NEWS.md"
authors: "geoMADE"
date: "12th April 2021"
output: html_document
---


Akmedoids' package updated (Version: v1.3.0)

Updates:
1. Added 4 new sample datasets, and used in the function examples. The datasets include `clustr`, `simulated`, `TO1Risk`, `traj_w_space`.
2. Added detailed descriptions of fields in each dataset. See `format` section of each dataset documentation.
3. Change the citation `Adepeju et al. 2019` to `Adepeju et al. 2021` based on the newly published article upon which the `akmedoids` package is based.
4. Modified the names of some functions, e.g. `akmedoids.clust` changed to `akclustr`,  `statPrint` changed to `print_akstats`, `population` changed to `popl`.
5. Added two new functions, namely; `remove_rows_n`, `plot_akstats` (see the documentation for details)
6. Ensured that the names of functions are in lower cases, e.g. `dataImputation` changed to `data_imputation`, `outlierDetect` changed to `outlier_detect`, etc.
7. Modify example in 'elbow_point' function to reduce run time


Your faithfully.
Monsuru.



R/data.R
#' Sample labels of cluster groups
#'
#' A dataframe of alphabetical labels representing
#' the optimal solution of `traj` dataset based on `akClust`
#' function
#'
#' @format A dataframe containing one variable:
#' \itemize{
#'   \item label: alphabetical label by clusters
#'     }
"clustr"


#' Simulated population data.
#'
#' Sample simulated population data to be used as
#' the denominator variable against `traj` dataset. Contains
#' data for two consecutive census years
#'
#' @format A dataframe with the following variables:
#' \itemize{
#'   \item location_id: Character id of sample census unit
#'   at which the population is obtained.
#'   \item census_2003: Population estimates at the sample
#'   locations for the census year 2003.
#'   \item census_2007: Population estimates at the sample
#'   locations for the census year 2007.
#'     }
"popl"

#' Simulated longitudinal dataset
#'
#' Contains simulated trajectories belonging to one of the
#' three pre-defined groups, namely (a) decreasing, (b) stable
#' and (c) increasing groups.
#'
#' @format A dataframe with the following variables:
#' \itemize{
#'   \item X1: Values across locations at time step 1
#'   \item X2: Values across locations at time step 2
#'   \item X3: Values across locations at time step 3
#'   \item X4: Values across locations at time step 4
#'   \item X5: Values across locations at time step 5
#'   \item X6: Values across locations at time step 6
#'   \item X7: Values across locations at time step 7
#'   \item X8: Values across locations at time step 8
#'   \item X9: Values across locations at time step 9
#'   \item X10: Values across locations at time step 10
#'   \item X11: Values across locations at time step 11
#'   \item X12: Values across locations at time step 12
#'   \item X13: Values across locations at time step 13
#'   \item X14: Values across locations at time step 14
#'   \item X15: Values across locations at time step 15
#'   \item X16: Values across locations at time step 16
#'   \item X17: Values across locations at time step 17
#'   \item X18: Values across locations at time step 18
#'   \item X19: Values across locations at time step 19
#'   \item X20: Values across locations at time step 20
#'   \item X21: Values across locations at time step 21
#'     }
"simulated"


#' Time-at-risk for the Adjudicated Toronto
#' Youth Data (Sample 1)
#'
#' Real-life time-at-risk per year for 378 individuals
#' from the age of 8 to 38 in the Toronto, Ontario, Canada.
#' The data is obtained through the R package `crimCV`.
#' For further information, please see: Nielsen, J. (2018)
#' crimCV: Group-Based Modelling of Longitudinal Data.
#' R package version 0.9.6.
#' URL https://CRAN.R-project.org/package=crimCV.
#'
#' @format A dataframe with the following variables:
#'   \itemize{
#'   \item 8: Time-at-risk per year at age 8
#'   \item 9: Time-at-risk per year at age 9
#'   \item 10: Time-at-risk per year at age 10
#'   \item 11: Time-at-risk per year at age 11
#'   \item 12: Time-at-risk per year at age 12
#'   \item 13: Time-at-risk per year at age 13
#'   \item 14: Time-at-risk per year at age 14
#'   \item 15: Time-at-risk per year at age 15
#'   \item 16: Time-at-risk per year at age 16
#'   \item 17: Time-at-risk per year at age 17
#'   \item 18: Time-at-risk per year at age 18
#'   \item 19: Time-at-risk per year at age 19
#'   \item 20: Time-at-risk per year at age 20
#'   \item 21: Time-at-risk per year at age 21
#'   \item 22: Time-at-risk per year at age 22
#'   \item 23: Time-at-risk per year at age 23
#'   \item 24: Time-at-risk per year at age 24
#'   \item 25: Time-at-risk per year at age 25
#'   \item 26: Time-at-risk per year at age 26
#'   \item 27: Time-at-risk per year at age 27
#'   \item 28: Time-at-risk per year at age 28
#'   \item 29: Time-at-risk per year at age 29
#'   \item 30: Time-at-risk per year at age 30
#'   \item 31: Time-at-risk per year at age 31
#'   \item 32: Time-at-risk per year at age 32
#'   \item 33: Time-at-risk per year at age 33
#'   \item 34: Time-at-risk per year at age 34
#'   \item 35: Time-at-risk per year at age 35
#'   \item 36: Time-at-risk per year at age 36
#'   \item 37: Time-at-risk per year at age 37
#'   \item 38: Time-at-risk per year at age 38
#'     }
"TO1Risk"

#' Sample longitudinal dataset
#'
#' Simulated longitudinal datasets containing
#' trajectories with missing values
#' (\code{NA}, \code{Inf}, \code{null})
#'
#' @format A dataframe with the following variables:
#'   \itemize{
#'   \item location_ids: Character id of sample locations
#'   at which values are obtained.
#'   \item X2001: Values at time step 1 (i.e. year 2001)
#'   \item X2002: Values at time step 2 (i.e. year 2002)
#'   \item X2003: Values at time step 3 (i.e. year 2003)
#'   \item X2004: Values at time step 4 (i.e. year 2004)
#'   \item X2005: Values at time step 5 (i.e. year 2005)
#'   \item X2006: Values at time step 6 (i.e. year 2006)
#'   \item X2007: Values at time step 7 (i.e. year 2007)
#'   \item X2008: Values at time step 8 (i.e. year 2008)
#'   \item X2009: Values at time step 9 (i.e. year 2009)
#'     }
"traj"

#' Sample longitudinal dataset containing whitespaces
#'
#' Longitudinal dataset with both trailing and leading
#' whitespaces. For example, there is a trailing whitespace
#' at cell [3, 6], while there is a leading whitespace
#' at cell [9, 4].
#'
#' @format A dataframe with the following variables:
#'   \itemize{
#'   \item location_ids: Character id of sample locations
#'   at which values are obtained.
#'   \item X2001: Values at time step 1 (i.e. year 2001)
#'   \item X2002: Values at time step 2 (i.e. year 2002)
#'   \item X2003: Values at time step 3 (i.e. year 2003)
#'   \item X2004: Values at time step 4 (i.e. year 2004)
#'   \item X2005: Values at time step 5 (i.e. year 2005)
#'   \item X2006: Values at time step 6 (i.e. year 2006)
#'   \item X2007: Values at time step 7 (i.e. year 2007)
#'   \item X2008: Values at time step 8 (i.e. year 2008)
#'   \item X2009: Values at time step 9 (i.e. year 2009)
#'     }
"traj_w_spaces"
